src/SelfManage/Verify/FallbackVerificationUsingOpenSsl.php

```diff
@@ -0,0 +1,297 @@
+             * verify the signature with the certificate.
+             *
+             *  - https://github.com/cli/cli/blob/234d2effd545fb9d72ea77aa648caa499aecaa6e/pkg/cmd/attestation/verify/verify.go#L225-L256
+             *  - https://docs.sigstore.dev/logging/verify-release/
+             *  - https://github.com/secure-systems-lab/dsse/blob/master/protocol.md#protocol
+             */
+            $this->assertCertificateClaims($attestation);
+            $output->writeln('#' . $attestationIndex . ': Certificate facts verified.', OutputInterface::VERBOSITY_VERBOSE);
+
+            $this->assertDigestFromAttestationMatchesActual($pharFilename, $attestation);
+            $output->writeln('#' . $attestationIndex . ': Payload digest matches downloaded file.', OutputInterface::VERBOSITY_VERBOSE);
+
+            $this->verifyDsseEnvelopeSignature($releaseMetadata, $attestationIndex, $attestation);
+            $output->writeln('#' . $attestationIndex . ': DSSE payload signature verified with certificate.', OutputInterface::VERBOSITY_VERBOSE);
+        }
+
+        $output->writeln('<info>✅ Verified the new PIE version (using fallback verification)</info>');
+    }
+
+    private function assertCertificateClaims(Attestation $attestation): void
+    {
+        $attestationCertificateInfo = openssl_x509_parse($attestation->certificate);
+        Assert::isArray($attestationCertificateInfo['extensions']);
+
+        /**
+         * See {@link https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#136141572641--fulcio} for details
+         * on the Fulcio extension keys; note the values are DER-encoded strings; the ASN.1 tag is UTF8String (0x0C).
+         *
+         * First up, check the extension values are what we expect; these are hard-coded, as we don't expect them
+         * to change unless the namespace/repo name change, etc.
+         */
+        foreach (self::ATTESTATION_CERTIFICATE_EXPECTED_EXTENSION_VALUES as $extension => $expectedValue) {
+            Assert::keyExists($attestationCertificateInfo['extensions'], $extension);
+            Assert::stringNotEmpty($attestationCertificateInfo['extensions'][$extension]);
+            $actualValue = $attestationCertificateInfo['extensions'][$extension];
+
+            // First character (the ASN.1 tag) is expected to be UTF8String (0x0C)
+            if (ord($actualValue[0]) !== 0x0C) {
+                throw FailedToVerifyRelease::fromMismatchingExtensionValues($extension, $expectedValue, $actualValue);
+            }
+
+            /**
+             * Second character is expected to be the length of the actual value
+             * as long as they are less than 127 bytes (short form)
+             *
+             * @link https://www.oss.com/asn1/resources/asn1-made-simple/asn1-quick-reference/basic-encoding-rules.html#Lengths
+             */
+            $derDecodedValue = substr($actualValue, 2, ord($actualValue[1]));
```

Comment:
Please add a check that the provided length actually matches the length of the contents. This is simple to do and prevents some potential shenanigans with the claimed length to be shorter than the length that OpenSSL extracts.

So basically:

```
"\x0c\x0512345678"
```

This claims to a length of 5, but the actual payload is 8 bytes in length. That should be rejected. And likewise should:

```
"\x0c\x05123"
```

I gave an example in https://github.com/php/pie/pull/205#discussion_r2022905578, but I guess you missed it, because I didn't mention it in the explanation (just in the suggested diff).

Reply:
Ah yes, I did miss that, thank you! :+1: